Renderer support for removing root components or updating parameters on them

SteveSandersonMS opened the pull request:
This is needed as part of #27574 

The core `Renderer` needs to have APIs for being told to:

 * Remove root components, clearing up properly and running all the correct disposal flows
 * Supply new parameters to root components, whether or not the rendering system is currently quiescent, and return a task that completes when we do reach quiescence

It's not necessary to add a new API for "supply new parameters" because the existing `RenderRootComponentAsync` actually does that already. The one thing it didn't support was being called while the system was not yet quiescent, so that's what I've added. It just involves juggling the tasks a little differently.

SteveSandersonMS committed:
Support removal of root components

SteveSandersonMS committed:
Support supplying new parameters to root components, even while not quiescent

javiercn commented on src/Components/Components/src/RenderTree/Renderer.cs:
```diff
@@ -199,53 +200,67 @@ protected Task RenderRootComponentAsync(int componentId)
+            // during a batch, but if a scenario emerges we can add support.
+            _batchBuilder.ComponentDisposalQueue.Enqueue(componentId);
+            _rootComponentsLatestParameters?.Remove(componentId);
+            ProcessRenderQueue();
```
Will this trigger the disposal of all the components in the tree? I remember having to do something more "sophisticated" here (triggering an empty render).

SteveSandersonMS replied:
It does - disposal is recursive. There's a unit test for that too.

javiercn replied:
Cool, I'm glad we found a simpler way, the way I did it was a bit convoluted

javiercn reviewed (commented):
Looks great!

I have to look at things with a fresh brain, but overall looks superb!

SteveSandersonMS commented on src/Components/Components/src/RenderTree/Renderer.cs:
```diff
@@ -254,21 +269,45 @@ protected async Task RenderRootComponentAsync(int componentId, ParameterView ini
+
+            try
+            {
+                _ongoingQuiescenceTask ??= ProcessAsynchronousWork();
```
```suggestion
                _ongoingQuiescenceTask = ProcessAsynchronousWork();
```

SteveSandersonMS commented on src/Components/Components/src/RenderTree/Renderer.cs:
```diff
@@ -254,21 +269,45 @@ protected async Task RenderRootComponentAsync(int componentId, ParameterView ini
         protected abstract void HandleException(Exception exception);
 
-        private async Task ProcessAsynchronousWork()
+        private Task? _ongoingQuiescenceTask;
```
TODO: Move this field to be with other fields

SteveSandersonMS committed:
Tidyings

SteveSandersonMS committed:
Fix unit tests, now we better validate use of sync context

Not regarding this as a breaking change, because it would never have been supported before to call the renderer from outside its sync context.

SteveSandersonMS committed:
Fix some unrelated unit tests while I'm here

SteveSandersonMS committed:
Update after rebase

SteveSandersonMS commented:
/azp run

SteveSandersonMS commented:
@javiercn Is there anything further you want to check here, or is this good to merge? I'm asking because I have a subsequent PR building on this and it would be handy to target `main` with that rather than chaining PRs.

javiercn commented on src/Components/Components/test/EventCallbackFactoryBinderExtensionsTest.cs:
```diff
@@ -493,7 +493,7 @@ public async Task CreateBinder_NullableDateTimeOffset()
 
             // Act
-            await binder.InvokeAsync(new ChangeEventArgs() { Value = expectedValue.ToString(CultureInfo.InvariantCulture), });
+            await binder.InvokeAsync(new ChangeEventArgs() { Value = expectedValue.ToString(CultureInfo.CurrentCulture), });
```
This seems to be unrelated, won't `CurrentCulture` cause issues with machines on other locales?

SteveSandersonMS replied:
The test was wrong before. Binding works with culture-specific values. The parsing was already culture-specific, so it was incorrect to test a culture-invariant input, and was failing on my machine (using months for days, and vice-versa).

The only reason it was passing in CI before was due to the cultural imperialism of `CultureInfo.InvariantCulture` being equivalent to US date formatting.

javiercn replied:
> the cultural imperialism of `CultureInfo.InvariantCulture` being equivalent to US date formatting

lol, fair enough :).

I only brought it up because we ran into issues in the past in this area.

javiercn reviewed (approved):
Looks great!
